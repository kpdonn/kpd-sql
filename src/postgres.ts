import { SqlPart, LookupParamNum } from "./everything"

export function printSql(sqlPart: SqlPart, pNum: LookupParamNum): string {
  const pr = (it: SqlPart): string => {
    switch (it.sqlKind) {
      case "plainJoin":
        return `join ${pr(it.joinTable)} on ${pr(it.onCondition)}`
      case "leftJoin":
        return `left join ${pr(it.joinTable)} on ${pr(it.onCondition)}`
      case "table":
        return `${it._table} ${it._tableAs}`
      case "hardcoded":
      case "placeholderParam":
        return `$${pNum(it) + 1}`
      case "and":
        return `(${pr(it.left)}) AND (${pr(it.right)})`
      case "or":
        return `(${pr(it.left)}) OR (${pr(it.right)})`
      case "column":
        return `${it._tableAs}.${it._column}`
      case "columnDeclaration":
        return `${pr(it.col)} as ${it.col._columnAs}`
      case "eq":
        return `${pr(it.left)} = ${pr(it.right)}`
      case "selectStatement":
        const columnsSql = it.columns.map(pr).join(", ")
        const fromSql = it.fromTables.map(pr).join(", ")
        const joinSql = it.joins.map(pr).join("\n")
        const whereSql = it.where ? `WHERE ${pr(it.where)}` : ""
        return `SELECT ${columnsSql} FROM ${fromSql} ${joinSql} ${whereSql}`
    }
  }
  return pr(sqlPart)
}
