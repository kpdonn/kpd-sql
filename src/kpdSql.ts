import * as t from "io-ts"
import {
  tblSym,
  colSym,
  tySym,
  col,
  ty,
  tblAsSym,
  tblAs,
  tbl,
  colAsSym,
  colAs,
} from "./implementation"
import { Table, ColInfo, Condition, Literal } from "./table"

export type ArrayKeys = keyof any[]

export type TupleKeys<T> = Exclude<keyof T, ArrayKeys>

export type TabCols<
  ColArr extends { [index: string]: ColInfo },
  ArrInd extends keyof ColArr,
  TblNames extends string
> = { [Ind in ArrInd]: ColArr[Ind][tblAsSym] extends TblNames ? Ind : never }[ArrInd]

export type Grab<
  ColArr extends { [index: string]: ColInfo },
  ArrInd extends keyof ColArr,
  TblNames extends string
> = [TabCols<ColArr, ArrInd, TblNames>] extends [never]
  ? {}
  : GetVal<ColArr[TabCols<ColArr, ArrInd, TblNames>][colAsSym], ColArr>

export type GetVal<
  ColNames extends string,
  ColArr extends { [index: string]: ColInfo }
> = { [K in ColNames]: FilterVal<K, ColArr[keyof ColArr]> }

export type TypeCol<N extends string, U extends t.Any> = { [colAs]: N; [ty]: U }

export type FilterVal<N extends string, T extends ColInfo> = T extends TypeCol<N, infer U>
  ? t.TypeOf<U>
  : never

export interface SQLFrom<
  Tables extends string = never,
  OptTables extends string = never,
  Cols = {},
  P = {}
> {
  from<T extends Table & NoDupTable<T, Tables>>(
    table: T
  ): SQLJoin<Tables | T[tblAsSym], OptTables, Cols, P>
}

export type GCol<C extends ColInfo> = C
export type SafeInd<T, K extends keyof Base, Base = T> = Extract<T, Base> extends never
  ? never
  : Extract<T, Base>[K]

export type NoDuplicates<T extends string> = { "NO DUPLICATE KEYS ALLOWED": T }
export type ColumnsObj<T> = { [K in TupleKeys<T>]: Extract<T[K], ColInfo> }

export interface SQLColumns<
  RT extends string,
  OT extends string,
  Cols = {},
  P = {},
  TblNames extends string = Literal<RT> | Literal<OT>
> {
  columns<
    C extends {
      [K in keyof T]: T[K][colAsSym] extends (
        | (SafeInd<T[Exclude<keyof T, K>], colAsSym>)
        | keyof Cols)
        ? NoDuplicates<SafeInd<T[K], colAsSym>>
        : ColInfo<TblNames>
    } & { "0": any },
    T extends ColumnsObj<C> = ColumnsObj<C>
  >(
    cols: C
  ): SQLSelectAndWhere<
    RT,
    OT,
    Cols &
      Grab<C, TupleKeys<C>, Literal<RT>> &
      EmptyIfNone<Partial<Grab<C, TupleKeys<C>, Literal<OT>>>>,
    P
  >
}

export type EmptyIfNone<T> = string extends keyof T
  ? {}
  : [keyof T] extends [never] ? {} : T

export interface SQLSelectAndWhere<RT extends string, OT extends string, Cols, P = {}>
  extends SQLColumns<RT, OT, Cols, P>,
    SQLWhere<RT, OT, Cols, P> {}

export type NoDupTable<T extends Table, TableNames extends string> = {
  [tbl]: T[tblAsSym] extends TableNames ? never : string
}
export interface SQLJoin<
  RT extends string,
  OT extends string,
  Cols,
  P = {},
  TblNames extends string = Literal<RT> | Literal<OT>
> extends SQLColumns<RT, OT, Cols, P> {
  join<T extends Table & NoDupTable<T, TblNames>, LCol extends ColInfo<TblNames>>(
    table: T,
    cond: Condition<TblNames | T[tblAsSym]>
  ): SQLJoin<RT | T[tblAsSym], OT, Cols, P>

  leftJoin<T extends Table & NoDupTable<T, TblNames>, LCol extends ColInfo<TblNames>>(
    table: T,
    cond: Condition<TblNames | T[tblAsSym]>
  ): SQLJoin<RT, OT | T[tblAsSym], Cols, P>
}

export type ExecuteArgs<P, Cols> = {} extends P
  ? () => Promise<Cols[]>
  : (args: P) => Promise<Cols[]>

export interface SQLExecute<RT extends string, OT extends string, Cols, P = {}>
  extends SQLReady<Cols> {
  execute: ExecuteArgs<P, Cols>
  toSql(): string
}

export declare const sqlReady: unique symbol
export interface SQLReady<Cols> {
  [sqlReady]: Cols
}

export interface SQLWhere<
  RT extends string,
  OT extends string,
  Cols,
  P = {},
  TblNames extends string = Literal<RT> | Literal<OT>
> extends SQLExecute<RT, OT, Cols, P> {
  where<CTbles extends TblNames, SP>(
    cond: Condition<CTbles, SP>
  ): SQLWhere<RT, OT, Cols, P & SP>

  whereSub<CTbles extends TblNames, SP>(
    cond: (subSelect: SQLFrom<RT, OT, {}>) => Condition<CTbles, SP>
  ): SQLWhere<RT, OT, Cols, P & SP>
}
