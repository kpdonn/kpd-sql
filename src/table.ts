import * as t from "io-ts"
import { ty, tbl, col, tySym, tblSym } from "./implementation"

export declare const condTbls: unique symbol
export type condTblsSym = typeof condTbls

export type InCol<CN extends string = string, Type extends t.Any = t.Mixed> = Record<
  CN,
  { type: Type }
>

export type ColInfo<
  TN extends string = string,
  Type extends t.Any = t.Mixed,
  CN extends string = string
> = {
  [ty]: Type
  [tbl]: TN
  [col]: CN
}

export type OutCol<
  TN extends string = string,
  Type extends t.Any = t.Mixed,
  CN extends string = string,
  ThisCol extends ColInfo = ColInfo<TN, Type, CN>
> = ThisCol & { as: AsCol<Type, TN> } & Comparisons<ThisCol>

export type AsCol<Type extends t.Any, TN extends string> = <NN extends string>(
  newName: NN
) => OutCol<TN, Type, NN>

export type TransformInCol<TN extends string, C extends InCol> = {
  [K in keyof C]: OutCol<TN, C[K]["type"], K>
}

export type Table<TN extends string = string> = {
  [tbl]: TN
}

export type TableOut<TN extends string, C extends InCol> = Table<TN> &
  TransformInCol<TN, C>

export function table<N extends string, C extends InCol>(arg: {
  name: N
  columns: C
}): TableOut<N, C> {
  const result: any = { [tbl]: arg.name }

  Object.keys(arg.columns).forEach(colName => {
    result[colName] = {
      [tbl]: arg.name,
      name: colName,
      ...arg.columns[colName]
    }
  })

  return result
}
export type Literal<T extends string> = string extends T ? never : T
export interface Condition<TblNames extends string> {
  [condTbls]: TblNames

  and<C extends Condition<any>>(cond: C): Condition<TblNames | C[condTblsSym]>
  or<C extends Condition<any>>(cond: C): Condition<TblNames | C[condTblsSym]>
}

export type Comparisons<Col1 extends ColInfo> = {
  eq: <Col2 extends ColInfo<any, Col1[tySym]>>(
    col2: Col2 | t.TypeOf<Col1[tySym]>
  ) => Condition<Col1[tblSym] | Literal<Col2[tblSym]>>
}

const Book = table({
  name: "book",
  columns: {
    id: { type: t.number },
    title: { type: t.string },
    year: { type: t.number },
    pages: { type: t.number },
    authorId: { type: t.number }
  }
})

const comp = Book.id.eq(Book.year)
