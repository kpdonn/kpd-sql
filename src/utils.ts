import * as t from "io-ts"
import { Table, Column, Condition } from "./table"

export const tbl: unique symbol = Symbol("tableName")
export const tblAs: unique symbol = Symbol("tableAs")
export const col: unique symbol = Symbol("columnName")
export const colAs: unique symbol = Symbol("columnAs")
export const ty: unique symbol = Symbol("typeSymbol")
export const condTbls: unique symbol = Symbol("condTbls")

export type tblSym = typeof tbl
export type colSym = typeof col
export type colAsSym = typeof colAs
export type tySym = typeof ty
export type tblAsSym = typeof tblAs
export type condTblsSym = typeof condTbls

export type Literal<T extends string> = string extends T ? never : T

export type SqlParamName<N extends string> = Record<"sqlParam", N>
export type SqlParam<N extends string, T> = string extends N ? {} : Record<N, T>

export declare const sqlReady: unique symbol
export interface SQLReady<Cols> {
  [sqlReady]: Cols
}

export type Comparisons<Col1Tbl extends string, Col1Type extends t.Any> = {
  not: Comparisons<Col1Tbl, Col1Type>

  eq<Col2 extends Column<any, Col1Type>, SPN extends string>(
    col2: Col2 | t.TypeOf<Col1Type> | SqlParamName<SPN>
  ): Condition<Col1Tbl | Literal<Col2[tblAsSym]>, SqlParam<SPN, t.TypeOf<Col1Type>>>

  isNull(): Condition<Col1Tbl>

  in(
    val: t.TypeOf<Col1Type>[] | SQLReady<Record<string, t.TypeOf<Col1Type>>>
  ): Condition<Col1Tbl>
}

export type InCol = { type: t.Any; dbName?: string }

export type LiteralOr<
  T extends undefined | string,
  D extends string
> = T extends undefined ? D : string extends T ? D : T

export type TransformInCol<TN extends string, C extends Record<string, InCol>> = {
  readonly [K in keyof C]: Column<TN, C[K]["type"], LiteralOr<C[K]["dbName"], K>, K>
}

export type TableWithColumns<
  TN extends string = string,
  C extends Record<string, InCol> = Record<string, InCol>,
  AsName extends string = TN
> = Table<TN, C, AsName> & TransformInCol<AsName, C>
