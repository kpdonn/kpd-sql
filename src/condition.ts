import { Literal, tblAsSym, ColType, tySym, SqlParam, SqlParamName } from "./utils"
import { Column } from "./table"

export const condTbls: unique symbol = Symbol("conditionTables")
export type cts = typeof condTbls
export const cp: unique symbol = Symbol("conditionParams")
export type cps = typeof cp

export type ConditionType = EqCondition | AndCondition | OrCondition

export abstract class Condition<CT extends string = string, P = {}> {
  readonly [condTbls]: CT;
  readonly [cp]: P
  abstract readonly kind: Literal<ConditionType["kind"]>

  and<C extends Condition>(cond: C): Condition<CT | C[cts], P & C[cps]> {
    return new AndCondition(this, cond)
  }
  or<C extends Condition>(cond: C): Condition<CT | C[cts], P & C[cps]> {
    return new OrCondition(this, cond)
  }
}

export class AndCondition<CT extends string = string, P = {}> extends Condition<CT, P> {
  readonly kind = "and"
  constructor(readonly left: Condition, readonly right: Condition) {
    super()
  }
}

export class OrCondition<CT extends string = string, P = {}> extends Condition<CT, P> {
  readonly kind = "or"

  constructor(readonly left: Condition, readonly right: Condition) {
    super()
  }
}

export class EqCondition<
  Col1 extends Column = Column,
  Col2 extends Column<string, Col1[tySym]> = never,
  SPN extends string = never
> extends Condition<
  Col1[tblAsSym] | Literal<Col2[tblAsSym]>,
  SqlParam<SPN, ColType<Col1>>
> {
  readonly kind = "eq"

  constructor(
    readonly left: Col1,
    readonly right: Col2 | ColType<Col1> | SqlParamName<SPN>
  ) {
    super()
  }
}
