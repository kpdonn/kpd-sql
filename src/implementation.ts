import * as Immutable from "immutable"
import { Pool } from "pg"
import { TableWithColumns } from "./utils"
import { Column } from "./table"
import { Condition } from "./condition"
import { SqlBuilder } from "./kpdSql"

let dbPool: Pool

export function initializePool(pool: Pool): void {
  dbPool = pool
}

interface Join {
  type: "inner" | "left"
  table: TableWithColumns
  cond: Condition<any>
}

const BuilderState = Immutable.Record({
  fromTable: (undefined as any) as TableWithColumns,
  joins: Immutable.List<Join>(),
  columns: Immutable.List(),
  wheres: Immutable.List(),
})
class SqlBuilderImpl {
  constructor(private readonly state = BuilderState()) {}

  from(table: TableWithColumns): SqlBuilderImpl {
    let newState = this.state.withMutations(arg => {
      arg.set("fromTable", table)
    })

    return new SqlBuilderImpl(newState)
  }

  join(table: TableWithColumns, cond: Condition<any>): SqlBuilderImpl {
    return this.addJoin("inner", table, cond)
  }

  leftJoin(table: TableWithColumns, cond: Condition<any>): SqlBuilderImpl {
    return this.addJoin("left", table, cond)
  }

  addJoin(
    type: "inner" | "left",
    table: TableWithColumns,
    cond: Condition<any>
  ): SqlBuilderImpl {
    const newState = this.state.updateIn(["joins"], joins =>
      joins.push({ type, table, cond })
    )

    return new SqlBuilderImpl(newState)
  }

  columns(cols: Column[]): SqlBuilderImpl {
    const newState = this.state.updateIn(["columns"], stateCols =>
      stateCols.withMutations((mut: any) => {
        cols.forEach(col => mut.push(col))
      })
    )
    return new SqlBuilderImpl(newState)
  }

  where(cond: Condition<any>): SqlBuilderImpl {
    const newState = this.state.updateIn(["wheres"], wheres => wheres.push({ cond }))

    return new SqlBuilderImpl(newState)
  }

  toSql(): string {
    let sql = "Select "
    const state = this.state
    const selectCols = state.columns.map(col => col.toSql()).join(", ")

    sql += selectCols + " \n"

    const fromClause = ` from ${state.fromTable} `

    sql += fromClause

    state.joins.forEach(join => {
      sql += " \n "
      const onClause = ` on ${join.cond} `
      sql += ` ${join.type} join ${join.table} ${onClause} `
    })

    sql += "\n where "

    state.wheres.forEach(where => {
      sql += ` ${where.cond.toSql()} `
    })

    return sql
  }

  execute(): Promise<any> {
    const sql = this.toSql()

    const result = dbPool.query(sql).then(res => res.rows)

    return result
  }
}

export function select(): SqlBuilder {
  return new SqlBuilderImpl() as any
}

export function tuple<T extends any[] & { "0": any }>(array: T): T {
  return array
}
