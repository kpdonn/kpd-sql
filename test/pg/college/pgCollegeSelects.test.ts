import { Pool } from "pg"
import { init, param } from "../../../src/everything"
import { PgPlugin } from "../../../src/postgres"
import { Class } from "./tables"
import * as fs from "fs"
import * as path from "path"

const connectionString = "postgresql://localhost:5432/sqltest"

const pool = new Pool({
  connectionString,
})

const db = init(PgPlugin.init(pool))

describe("college select tests", () => {
  it("empty test", async () => {
    const query = db
      .select()
      .from(Class)
      .groupBy([Class.id, Class.courseId])
      .columns([Class.semesterId])
  })

  it("group by with test", async () => {
    const query = db
      .with(
        "myWith",
        db
          .from(Class)
          .columns([Class.id, Class.courseId, Class.semesterId, Class.professorId])
      )
      .select(sq =>
        sq
          .from(sq.table.myWith)
          .groupBy([sq.table.myWith.id])
          .columns([sq.table.myWith.id, sq.table.myWith.courseId])
      )
  })
})

beforeAll(async () => {
  const initDb = fs.readFileSync(
    path.resolve(__dirname, "postgres-college-init.sql"),
    "utf8"
  )

  await pool.query(initDb)
})

afterAll(async () => {
  await pool.end()
})
