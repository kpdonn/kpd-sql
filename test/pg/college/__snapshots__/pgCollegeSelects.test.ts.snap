// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`college select tests Students credit hours by semester 1`] = `
"SELECT
\\"student\\".\\"id\\" as \\"id\\",
\\"student\\".\\"first_name\\" as \\"firstName\\",
\\"student\\".\\"last_name\\" as \\"lastName\\",
\\"student\\".\\"major_id\\" as \\"majorId\\",
\\"class\\".\\"semester_id\\" as \\"semesterId\\",
sum(\\"course\\".\\"credit_hours\\") as \\"sum\\"
FROM
( ( ( \\"student\\" JOIN \\"student_class\\" ON \\"student\\".\\"id\\" = \\"student_class\\".\\"student_id\\" ) JOIN \\"class\\" ON \\"class\\".\\"id\\" = \\"student_class\\".\\"class_id\\" ) JOIN \\"course\\" ON \\"course\\".\\"id\\" = \\"class\\".\\"course_id\\" )
GROUP BY \\"student\\".\\"id\\", \\"class\\".\\"semester_id\\""
`;

exports[`college select tests Students credit hours by semester 2`] = `
Array [
  anonymous {
    "firstName": "Davonte",
    "id": 1,
    "lastName": "Leannon",
    "majorId": null,
    "semesterId": 1,
    "sum": "7",
  },
  anonymous {
    "firstName": "Lester",
    "id": 13,
    "lastName": "Wuckert",
    "majorId": 2,
    "semesterId": 2,
    "sum": "11",
  },
  anonymous {
    "firstName": "Raven",
    "id": 11,
    "lastName": "Lebsack",
    "majorId": 2,
    "semesterId": 1,
    "sum": "3",
  },
  anonymous {
    "firstName": "Arno",
    "id": 18,
    "lastName": "Bradtke",
    "majorId": 3,
    "semesterId": 2,
    "sum": "22",
  },
  anonymous {
    "firstName": "Murl",
    "id": 3,
    "lastName": "Kuphal",
    "majorId": null,
    "semesterId": 2,
    "sum": "18",
  },
  anonymous {
    "firstName": "Grady",
    "id": 6,
    "lastName": "Rau",
    "majorId": 1,
    "semesterId": 1,
    "sum": "8",
  },
  anonymous {
    "firstName": "Murl",
    "id": 3,
    "lastName": "Kuphal",
    "majorId": null,
    "semesterId": 1,
    "sum": "5",
  },
  anonymous {
    "firstName": "Norris",
    "id": 8,
    "lastName": "Johnson",
    "majorId": 1,
    "semesterId": 1,
    "sum": "9",
  },
  anonymous {
    "firstName": "Herminio",
    "id": 12,
    "lastName": "Morissette",
    "majorId": 2,
    "semesterId": 1,
    "sum": "6",
  },
  anonymous {
    "firstName": "Imani",
    "id": 14,
    "lastName": "Kuhlman",
    "majorId": 2,
    "semesterId": 2,
    "sum": "20",
  },
  anonymous {
    "firstName": "Cristobal",
    "id": 7,
    "lastName": "Wyman",
    "majorId": 1,
    "semesterId": 1,
    "sum": "9",
  },
  anonymous {
    "firstName": "Timothy",
    "id": 5,
    "lastName": "Metz",
    "majorId": null,
    "semesterId": 1,
    "sum": "15",
  },
  anonymous {
    "firstName": "Timothy",
    "id": 5,
    "lastName": "Metz",
    "majorId": null,
    "semesterId": 2,
    "sum": "8",
  },
  anonymous {
    "firstName": "Lester",
    "id": 13,
    "lastName": "Wuckert",
    "majorId": 2,
    "semesterId": 1,
    "sum": "9",
  },
  anonymous {
    "firstName": "Cristobal",
    "id": 7,
    "lastName": "Wyman",
    "majorId": 1,
    "semesterId": 2,
    "sum": "27",
  },
  anonymous {
    "firstName": "Norris",
    "id": 8,
    "lastName": "Johnson",
    "majorId": 1,
    "semesterId": 2,
    "sum": "20",
  },
  anonymous {
    "firstName": "Georgette",
    "id": 17,
    "lastName": "Fisher",
    "majorId": 3,
    "semesterId": 2,
    "sum": "10",
  },
  anonymous {
    "firstName": "Charlie",
    "id": 16,
    "lastName": "Bashirian",
    "majorId": 3,
    "semesterId": 2,
    "sum": "21",
  },
  anonymous {
    "firstName": "Grady",
    "id": 6,
    "lastName": "Rau",
    "majorId": 1,
    "semesterId": 2,
    "sum": "14",
  },
  anonymous {
    "firstName": "Hosea",
    "id": 19,
    "lastName": "Ziemann",
    "majorId": 3,
    "semesterId": 2,
    "sum": "25",
  },
  anonymous {
    "firstName": "Alexandria",
    "id": 4,
    "lastName": "Reinger",
    "majorId": null,
    "semesterId": 2,
    "sum": "5",
  },
  anonymous {
    "firstName": "Hosea",
    "id": 19,
    "lastName": "Ziemann",
    "majorId": 3,
    "semesterId": 1,
    "sum": "4",
  },
  anonymous {
    "firstName": "Queenie",
    "id": 9,
    "lastName": "Baumbach",
    "majorId": 1,
    "semesterId": 1,
    "sum": "16",
  },
  anonymous {
    "firstName": "Davonte",
    "id": 1,
    "lastName": "Leannon",
    "majorId": null,
    "semesterId": 2,
    "sum": "4",
  },
  anonymous {
    "firstName": "Georgette",
    "id": 17,
    "lastName": "Fisher",
    "majorId": 3,
    "semesterId": 1,
    "sum": "15",
  },
  anonymous {
    "firstName": "Trisha",
    "id": 2,
    "lastName": "Kertzmann",
    "majorId": null,
    "semesterId": 2,
    "sum": "7",
  },
  anonymous {
    "firstName": "Trisha",
    "id": 2,
    "lastName": "Kertzmann",
    "majorId": null,
    "semesterId": 1,
    "sum": "13",
  },
  anonymous {
    "firstName": "Alexandria",
    "id": 4,
    "lastName": "Reinger",
    "majorId": null,
    "semesterId": 1,
    "sum": "11",
  },
  anonymous {
    "firstName": "Arno",
    "id": 18,
    "lastName": "Bradtke",
    "majorId": 3,
    "semesterId": 1,
    "sum": "6",
  },
  anonymous {
    "firstName": "Kariane",
    "id": 10,
    "lastName": "Gleichner",
    "majorId": 1,
    "semesterId": 2,
    "sum": "15",
  },
  anonymous {
    "firstName": "Herminio",
    "id": 12,
    "lastName": "Morissette",
    "majorId": 2,
    "semesterId": 2,
    "sum": "14",
  },
  anonymous {
    "firstName": "Queenie",
    "id": 9,
    "lastName": "Baumbach",
    "majorId": 1,
    "semesterId": 2,
    "sum": "5",
  },
  anonymous {
    "firstName": "Raven",
    "id": 11,
    "lastName": "Lebsack",
    "majorId": 2,
    "semesterId": 2,
    "sum": "1",
  },
  anonymous {
    "firstName": "Tianna",
    "id": 15,
    "lastName": "Walter",
    "majorId": 2,
    "semesterId": 1,
    "sum": "10",
  },
  anonymous {
    "firstName": "Tianna",
    "id": 15,
    "lastName": "Walter",
    "majorId": 2,
    "semesterId": 2,
    "sum": "15",
  },
  anonymous {
    "firstName": "Charlie",
    "id": 16,
    "lastName": "Bashirian",
    "majorId": 3,
    "semesterId": 1,
    "sum": "2",
  },
  anonymous {
    "firstName": "Winona",
    "id": 20,
    "lastName": "Raynor",
    "majorId": 3,
    "semesterId": 1,
    "sum": "38",
  },
  anonymous {
    "firstName": "Imani",
    "id": 14,
    "lastName": "Kuhlman",
    "majorId": 2,
    "semesterId": 1,
    "sum": "6",
  },
]
`;

exports[`college select tests number students 1`] = `
"SELECT
count(*) as \\"count\\"
FROM
\\"student\\""
`;

exports[`college select tests number students 2`] = `
Array [
  anonymous {
    "count": "20",
  },
]
`;

exports[`college select tests trivial group by test 1`] = `
"SELECT
\\"class\\".\\"semester_id\\" as \\"semesterId\\"
FROM
\\"class\\"
GROUP BY \\"class\\".\\"id\\", \\"class\\".\\"course_id\\""
`;

exports[`college select tests trivial group by test 2`] = `
Array [
  anonymous {
    "semesterId": 2,
  },
  anonymous {
    "semesterId": 1,
  },
  anonymous {
    "semesterId": 1,
  },
  anonymous {
    "semesterId": 2,
  },
  anonymous {
    "semesterId": 2,
  },
  anonymous {
    "semesterId": 1,
  },
  anonymous {
    "semesterId": 2,
  },
  anonymous {
    "semesterId": 2,
  },
  anonymous {
    "semesterId": 1,
  },
  anonymous {
    "semesterId": 2,
  },
  anonymous {
    "semesterId": 1,
  },
  anonymous {
    "semesterId": 1,
  },
  anonymous {
    "semesterId": 1,
  },
  anonymous {
    "semesterId": 2,
  },
  anonymous {
    "semesterId": 1,
  },
  anonymous {
    "semesterId": 2,
  },
  anonymous {
    "semesterId": 2,
  },
  anonymous {
    "semesterId": 1,
  },
  anonymous {
    "semesterId": 2,
  },
  anonymous {
    "semesterId": 2,
  },
  anonymous {
    "semesterId": 2,
  },
  anonymous {
    "semesterId": 2,
  },
  anonymous {
    "semesterId": 2,
  },
  anonymous {
    "semesterId": 2,
  },
  anonymous {
    "semesterId": 2,
  },
  anonymous {
    "semesterId": 2,
  },
  anonymous {
    "semesterId": 1,
  },
  anonymous {
    "semesterId": 2,
  },
  anonymous {
    "semesterId": 1,
  },
  anonymous {
    "semesterId": 2,
  },
  anonymous {
    "semesterId": 1,
  },
  anonymous {
    "semesterId": 2,
  },
  anonymous {
    "semesterId": 2,
  },
  anonymous {
    "semesterId": 1,
  },
  anonymous {
    "semesterId": 1,
  },
  anonymous {
    "semesterId": 1,
  },
]
`;
